### ISSUE TITLE
In metrics.calculate, where clauses on columns not present in dimensions do not work.

### ISSUE THREAD

**Issue by hsenger-tiger** [dbt-labs/dbt_metrics#228]
### Is this a new bug in dbt_metrics?

- [X] I believe this is a new bug in dbt_metrics
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

If I have a metric defined like so:

metrics:
  - name: transactions
    label: Transactions
    model: ref('base_table')
    calculation_method: count_distinct
    expression: transaction_id
    timestamp: date
    time_grains: [day, week, month]
    dimensions:
      - company
   
And then I do:

select *
from {{ metrics.calculate(
    metric('transactions'),
    grain='month',
    where="company in ('Company A', 'Company B')"
) }}

It fails with the following SQL compilation error:

invalid identifier 'COMPANY'

The only way to get this error to go away is to add dimensions = ['company'] to the metrics.calculate macro, but then it won't aggregate across companies like I want it to.

### Expected Behavior

From the DBT documentation (and to support basic things like aggregating across multiple dimension values), I would expect 

select *
from {{ metrics.calculate(
    metric('transactions'),
    grain='month',
    where="company in ('Company A', 'Company B')"
) }}

to not fail to compile, and instead work and compute the specified metrics for records in the base table that meet the where condition.

### Steps To Reproduce

1. Create a metric with the definition like the above.
2. Create a DBT model that does a metrics.calculate with a where clause on a dimension field not present in the metrics.calculat.e
3. Observe SQL compilation failure.

### Environment

```markdown
- dbt-adapter & version: dbt-snowflake v1.4.2
- dbt_metrics version: v1.4.1
```

### Which database adapter are you using with dbt?

snowflake

**Comment by callum-mcdata** [comment]
Hi @hsenger-tiger . This is a known limitation of the `where` clause in the macro because it only applies to the final CTE.

Here is the documentation on this:

> A sql statement, or series of sql statements, that alter the final CTE in the generated sql. Most often used to limit the data to specific values of dimensions provided

There are currently no plans to change the `where` functionality to apply filters in the base sql query. That being said, this is something we're hoping to address in the MetricFlow integration!

**Comment by hsenger-tiger** [comment]
It looks like this used to work at some point:

https://github.com/dbt-labs/dbt_metrics/pull/34

Then it looks like folks decided to make it not work here:

https://github.com/dbt-labs/dbt_metrics/pull/69

because of a philosophical concern. If this is something you're hoping to address / change with the MetricFlow integration, I'm curious why not just undo the hotfix that broke this in the interim to help folks?

**Comment by callum-mcdata** [comment]
@hsenger-tiger the initial implementation was actually by me! It was an incorrect implementation in that I misunderstood the goal - it was my third week at dbt Labs when I pushed that one 😆 . We made the hotfix to have it be more in line with what was suggested. 

Unfortunately we're not willing to revert that hotfix as the package has evolved since then and would not fit within the new sql generation capabilities. Especially things like multiple metrics that come from different tables - we don't parse the syntax as SQL so we can't know exactly where it is coming from. 

Good callout on the docs though! I just pushed through that fix.

**Comment by hsenger-tiger** [comment]
Thanks Callum. What's the recommendation for how to enable folks to compute metrics across multiple values of a dimension? It's a very common use case at our organization, and the set of values they want to filter to / aggregate over is not predictable and there are 100s of permutations folks may want it for. It seems like a natural use case to want to compute a metric over a set of dimensions rather than just 1. We could ask them to compute for each dimension and then aggregate separately, but for things like count distinct that does not work.

**Comment by callum-mcdata** [comment]
I think this is where we point at the value of the Semantic Layer offering - this package is really built to support a dynamic querying use case instead of materializing across the 100s of permutations. We have been fairly clear in the community that materializing metrics into models is not the ideal pattern that we are aiming for. 

The reason is exactly to your point - you run into the combinatorial problem with a large number of dimensions and grains and its further complicated with metric types that are semi or non-additive like count distinct.

**Comment by hsenger-tiger** [comment]
Thanks Callum. Apologies if this is a silly question... but how would I use the Semantic Layer to compute a metric across multiple dimension values? We are using the Semantic Layer API, and it has the same limitation from what I can tell.

**Comment by callum-mcdata** [comment]
@hsenger-tiger we may be speaking around one another right now - lets take a step back. What exactly do you mean:

> compute a metric across multiple dimension values

Can you provide some examples here? In my head that means calculating a metric with n+1 dimensions, like revenue by location & customer name. 

> We are using the Semantic Layer API

We don't have a Semantic Layer API so I think there might be some confusion here. The currently released implementation of the Semantic Layer connects by proxying a Snowflake JDBC from your tool of choice. But to your point, yes it is using the same SQL compilation.

**Comment by hsenger-tiger** [comment]
Yes, let me clarify. It is not that we want to calculate a metric by 2 different dimensions like by location and customer name, but rather that we want to calculate a metric for multiple values of the same dimension. For example, calculate revenue across 2 locations. So for example if we have a location field as a dimension for our metric, I would like to be able to calculate that metric not just for 1 location at a time always, but also for several locations.

The example in the initial bug report is a bit clearer here, but here's another in case it helps. Given a metric definition for a metric named "revenue" with a dimension field called "country":

name: revenue
label: Revenue
model: ref('base_table')
calculation_method: sum
expression: revenue
timestamp: date
time_grains: [day, week, month]
dimensions:
- country

I would like to be able to do things like this (either via metrics.calculate in a DBT model, or in a metrics.calculate passed to the DBT proxy server on-the-fly):

{{ metrics.calculate(
metric('revenue'),
grain='month',
where="country in ('US', 'CA')"
) }}

To get monthly revenue for 2 countries over time. Currently this is not possible, since to use country in the where, I also need to specify it in the dimensions in the calculate call, which will give me the revenue in each country, not the aggregated revenue across the 2 countries. For metrics that are simple sums, this is bit annoying / confusing but easy to work around. For metrics that use a count_distinct like unique customers however, it's a much larger issue.

**Comment by callum-mcdata** [comment]
Ahhh got it. I appreciate the clarification with the example. Unfortunately this behavior is not supported today and probably won't be in the near future as the team is focusing on the MetricFlow integration. 

If this is something that you frequently look at, I'd currently recommend pushing this logic down to the logical layer (ie the model) and creating some flag there. Then create a new metric called something like `us_revenue` where the filter is set for that flag to be true. Obviously it's a bit more configuration than ideal but does help ensure that the logic to calculate the thing is in your repo!
